alg/teca_regional_moisture_flux.cxx

```diff
@@ -0,0 +1,690 @@
+    const std::vector<teca_metadata> &input_md)
+{
+#ifdef TECA_DEBUG
+    std::cerr << teca_parallel_id()
+        << "teca_regional_moisture_flux::get_output_metadata" << std::endl;
+#endif
+    (void)port;
+
+    // check that the class has been configured correctly
+    if (this->ivt_u_variable.empty())
+    {
+        TECA_FATAL_ERROR("The ivt_u_variable was not set")
+        return teca_metadata();
+    }
+
+    if (this->ivt_v_variable.empty())
+    {
+        TECA_FATAL_ERROR("The ivt_u_variable was not set")
+        return teca_metadata();
+    }
+
+    if (this->region_mask_variable.empty())
+    {
+        TECA_FATAL_ERROR("The region_mask_variable was not set")
+        return teca_metadata();
+    }
+
+
+    if (this->moisture_flux_variable.empty())
+    {
+        TECA_FATAL_ERROR("The moisture_fux_variable was not set")
+        return {};
+    }
+
+    // add in the array we will generate
+    teca_metadata out_md(input_md[0]);
+    out_md.append("variables", this->moisture_flux_variable);
+
+    // add attributes for the netcdf writer
+    teca_metadata attributes;
+    out_md.get("attributes", attributes);
+
+    if (this->moisture_flux_attributes.empty())
+    {
+        // generate and cache the attributes. they're also needed in execute
+        std::ostringstream descr;
+        descr << "moisture flux computed from IVT vector (" << this->ivt_u_variable
+            << ", " << this->ivt_v_variable << ") over a region defined by the mask "
+            << this->region_mask_variable;
+
+        teca_array_attributes atts{teca_variant_array_code<double>::get(),
+            teca_array_attributes::no_centering, 0, teca_array_attributes::none_active(),
+            "kg s-1", this->moisture_flux_variable, descr.str(), 1, 1.e20};
+
+        attributes.append(this->moisture_flux_variable, (teca_metadata)atts);
+
+        this->moisture_flux_attributes = (teca_metadata)atts;
+    }
+    else
+    {
+        // use cached attributes
+        attributes.append(this->moisture_flux_variable,
+            this->moisture_flux_attributes);
+    }
+
+    out_md.set("attributes", attributes);
+
+    return out_md;
+}
+
+// --------------------------------------------------------------------------
+std::vector<teca_metadata> teca_regional_moisture_flux::get_upstream_request(
+    unsigned int port, const std::vector<teca_metadata> &input_md,
+    const teca_metadata &request)
+{
+    (void)port;
+    (void)input_md;
+
+    std::vector<teca_metadata> up_reqs;
+
+    // copy the incoming request to preserve the downstream requirements and
+    // add the arrays we need
+    teca_metadata req(request);
+
+    std::set<std::string> arrays;
+    if (req.has("arrays"))
+        req.get("arrays", arrays);
+
+    arrays.insert(this->ivt_u_variable);
+    arrays.insert(this->ivt_v_variable);
+    arrays.insert(this->region_mask_variable);
+
+    // TODO if the valid value masks for IVT are available request them too
```

Comment:
flagging this TODO in a comment so it doesn't get missed

Reply:
@taobrienlbl Question, do we need to handle missing values in IVT field? I'm guessing it would be a good idea, but maybe save it for future work?

Reply:
Hmm. Might make sense to rename this algorithm to `teca_atmospheric_flux` and use some generic names like vector_u and vector_v instead of ivt_u and v?

Reply:
Oh wait, I think we've made some simplifications in our derivation that only apply to IVT. So I guess that's not going to work.